include/circt/Dialect/FIRRTL/FIRRTLStatements.td

```diff
@@ -184,3 +184,12 @@ def WhenOp : FIRRTLOp<"when", [SingleBlock, NoTerminator, NoRegionArguments,
     }
   }];
 }
+
+def ForceOp : FIRRTLOp<"force", [SameTypeOperands]> {
+  let summary = "Force procedural statement";
+  let description = "Maps to the corresponding `sv.force` operation.";
+  let arguments = (ins FIRRTLType:$dest, FIRRTLType:$src);
+  let results = (outs);
+  let assemblyFormat =
+    "$dest `,` $src attr-dict `:` type($dest) `,` type($src)";
+}
```

Comment:
It would be cool if there was a way to cast a FIRRTL type to an SV type and then use SV force statements inside FIRRTL modules.  An alternative approach, an perhaps a better one, would be to just generate SV modules with SV ops (e.g., `force`) in them.  What do you think?

As we talked about, I get nervous with an explosion of "higher level dialects" needing to replicate the ops of "lower level dialects". I'm not sure of an awesome solution here, though.

Reply:
Yeah you have a good point here. The thing that is difficult here is that the pass must run before `LowerTypes` (because we don't support splitting annotations on aggregates apart into multiple annotations on the individual fields). But that then requires the generated IR operations to be properly handled by `LowerTypes`.

Right now with the `ForceOp` that should be handled properly, and insert adequate subfield accesses on the verbatim XMRs so they lower properly. Not sure how we would do that if there were SV ops in the game already. We'd have to have some scheme for lowering generic SV ops in `LowerTypes`, e.g. by replicating operations on an aggregate to multiple instances operating on the individual fields.

Reply:
It is possible to do this today(have mixed SV and FIRRTL dialects inside a firrtl.module), for integer values with the FIRRTL dialect `StdIntCastOp` operation (structs and analog/inout have comparable ops).

I think the proposed patch is a logical step forward... but if this is the first of many weird SV things we have to pull forward, then I agree with seldridge that we should investigate mixing dialects.

Reply:
Agreed, and I could totally see us want more and more SV ops. However, making sure these run through `LowerTypes` in a predictable manner might be tricky. The alternative would be to make the annotations run through `LowerTypes` properly, which might be equally tricky :thinking: Not sure what the best solution is here. Would it make sense to hold off on dialect mixing until we have a bunch more use cases and then try to tackle those all in one go?

Reply:
I'd just only generate integer type ops (not structs etc), then LowerTypes would just not touch them at all.

Reply:
I don't think that would work in this case, because the annotations operate on aggregate types. Well, this pass could just generate a bunch of subfield/subindex ops and emit multiple `ForceOp`s to drive the individual fields. But I'm not sure we'd want to have passes each roll their own limited version of type lowering whenever we want to use an SV op. Sounds like an awful lot of replication.

You could also argue that since 80% of FIRRTL semantics is in the annotations, the signal forcing that this annotation triggers should be represented as an op in the dialect. That would probably hold for quite a few other annotations, like wiring/boring utilities, data/mem taps, Schuyler's interfaces, etc.. But then again, these passes are like one-off collateral generation thingies, so introducing separate ops and being able to reason about the output in its entirety might by overkill.

Reply:
Ah, I didn’t realize it was for aggregates as well.  I understand the lower types concerns now.  I’m good with adding these to the firrtl dialect, thanks!

Reply:
While the annotation is defined on aggregate types, does it have to be resolved on aggregate types? The whole Grand Central pass stack in the SFC is running on low FIRRTL and I don't know if we need to deviate here?

Reply:
Or: getting annotations to run through lower types can be tricky, but is usually pretty straightforward if they're scattered properly. I.e., if the scattering is placing trackers on aggregates then the trackers are copied to the blown out type. Alternatively, individual trackers can be placed on each leaf field/index and lower types will place these correctly on the lowered type.

Specializing lower types to handle weird, compound annotations is a bad idea, though.

Reply:
True, we could just say we run after `LowerTypes`. For stuff like these source/sink connect annotations, `LowerTypes` would probably also have to attach a field ID to the replicated annotation. Otherwise we'd run into trouble identifying which field pairs actually match up.

I'm starting to like that idea. Scattering in a way such that the annotations survive LowerTypes without any modifications is great, and then the pass can just assume there are no aggregates. Pushing the annos through LowerTypes was what scared me. Not sure how we would do that. Do we add a list of annotations to LowerTypes that need to have a field ID added when they're pushed down to lowered values?

Reply:
Either the scattering can scatter to leaves or lower types may need to add an "original order" tag to annotations that it lowers. Probably doing the latter is better because it means scattering doesn't need to know how to lower.

Reply:
The scattering now takes care of this and `LowerTypes` adds a tag to the annotations as they get replicated during lowering.